File: python/kohaku/src/kohaku/composer/fallback.py
"""Deterministic fallback Spec (port of TS fallback.ts).

When L1 is exhausted and L2 is not permitted, display a summary of the Intent using only presentMarkdown,
without using the LLM at all.
"""

from __future__ import annotations

from typing import Literal

from kohaku.spec import SPEC_VERSION, Intent, UISpec, canonical_stringify


def build_fallback_spec(
    *,
    intent: Intent,
    data_version: str,
    reason: str,
    composed_by: str,
    cache: Literal["hit", "miss", "bypass", "fixated"] = "miss",
    from_tier: Literal["L1", "L2"] = "L1",
) -> UISpec:
    """from_tier is the tier that actually failed. It is reflected in both provenance.tier and fallback.from."""
    markdown = "\n".join(
        [
            "### Could not render this request",
            "",
            f"Requested intent: `{intent.canonical}`",
            "",
            "```json",
            canonical_stringify(intent.params),
            "```",
            "",
            f"Reason: {reason}",
        ]
    )
    return UISpec.model_validate(
        {
            "kohaku": SPEC_VERSION,
            "intent": intent.to_wire(),
            "dataVersion": data_version,
            "components": [
                {
                    "id": "root",
                    "type": "layout.stack",
                    "props": {"direction": "vertical", "gap": "md"},
                    "children": ["md1"],
                },
                {"id": "md1", "type": "presentMarkdown", "props": {"markdown": markdown}},
            ],
            "events": [],
            "provenance": {
                "tier": from_tier,
                "composedBy": composed_by,
                "cache": cache,
                "fallback": {"from": from_tier, "reason": reason, "kind": "generation"},
            },
        }
    )

File: packages/composer/src/fallback.ts
import { type CanonicalIntent, canonicalStringify, SPEC_VERSION, type UISpec } from "@kohaku-ui/spec-core";

/**
 * Deterministic fallback Spec used when L1 is exhausted and L2 is not permitted either.
 * Uses no LLM at all, showing a summary of the Intent with presentMarkdown only.
 */
export function buildFallbackSpec(args: {
  intent: CanonicalIntent;
  dataVersion: string;
  reason: string;
  composedBy: string;
  /** Observation label per cacheMode (do not mislabel a fallback under bypass as miss) */
  cache?: UISpec["provenance"]["cache"];
  /**
   * The tier that actually failed. Reflected in both provenance.tier and fallback.from.
   * When L2 fails via routeTier=L2 direct entry or an L1→L2 promotion, passing "L2" prevents
   * mislabeling (which would otherwise always be "L1").
   * Unspecified defaults to "L1" (backward compatible).
   */
  from?: "L1" | "L2";
}): UISpec {
  const from = args.from ?? "L1";
  return {
    kohaku: SPEC_VERSION,
    intent: args.intent,
    dataVersion: args.dataVersion,
    components: [
      {
        id: "root",
        type: "layout.stack",
        props: { direction: "vertical", gap: "md" },
        children: ["md1"],
      },
      {
        id: "md1",
        type: "presentMarkdown",
        props: {
          markdown: [
            "### Could not render this request",
            "",
            `Requested intent: \`${args.intent.canonical}\``,
            "",
            "```json",
            canonicalStringify(args.intent.params),
            "```",
            "",
            `Reason: ${args.reason}`,
          ].join("\n"),
        },
      },
    ],
    events: [],
    provenance: {
      tier: from,
      composedBy: args.composedBy,
      cache: args.cache ?? "miss",
      fallback: { from, reason: args.reason, kind: "generation" },
    },
  };
}
